### ISSUE TITLE
Actual usage of reconstruction loss on initialization stage

### ISSUE THREAD

**Issue by the-jb** [sunnweiwei/GenRet#4]
Dear authors.

I was very impressed with your amazing work, which enables training dynamic document clustering through tokenization process.
I have carefully read your paper and the code to understand the details correctly but still have a question about the intent of reconstruction loss on initialization stage.

First, in your paper, it is said for building reconstruction loss $\mathcal L_{Rec}$ as:
> warm-up the model by passing the continuous representation $d_T$ to the reconstruction model instead of the docid
representation $z_T$

But, I found that the code actually calculates reconstruction loss as below, and I cannot find any replacement described above in the code:
https://github.com/sunnweiwei/GenRet/blob/d3c16091f853a6a3b810d5cbf00128625ef39b90/run.py#L403-L404

Furthermore, in `main()` function, the `loss_w` is configured as `1`:
https://github.com/sunnweiwei/GenRet/blob/d3c16091f853a6a3b810d5cbf00128625ef39b90/run.py#L1154 , which is the pre-configured weight that exclude the reconstruction loss as below.
https://github.com/sunnweiwei/GenRet/blob/d3c16091f853a6a3b810d5cbf00128625ef39b90/run.py#L687-L688

In summary, I found some inconsistency between the paper and the code in calculating reconstruction loss on codebook initialization stage.
If I understand correctly, the reconstruction is actually not being used on the initialization phase, is this correct?

Thanks for sharing your wonderful work.
Best regards,

JB

**Comment by sunnweiwei** [comment]
Hi,

Thank you for pointing out this issue. The code is inconsistent with the original paper in this part. Some analysis shows that with the query-doc contrastive loss, the model outputs document embeddings used for initialization; and the doc-doc contrastive loss (reconstruction loss) may be less useful and is omitted.

Best, Weiwei

**Comment by the-jb** [comment]
Thanks for your reply, @sunnweiwei.
Regarding to your answer, I meet another question.

As you mentioned, the code uses query-doc contrastive loss on initialization phase, which seems to be the *Retrieval loss* of your paper.
If reconstruction loss doesn't play a significant role, I am wondering how do you think about the constrained k-means clustering when initializing codebook.
I am asking because it seems there having not much difference between initialization and training except that the centroids are initialized with the k-means clustering before training.
Since Sinkhorn-Knopp iterations balances the `docid` assignment on training, I was wondering why you have used both balacing algorithms for dealing with the codebook.
